text/0161-telemetry-schema-and-resource.md

```diff
@@ -0,0 +1,196 @@
+}
+```
+
+### Understanding schema compatibility
+
+The second will be done via a new (public) SDK method that allows understanding
+whether or not two schemas are compatible, and which is the "newer" version.
+
+e.g. in Java you would have:
+
+```java
+public abstract class Schemas {
+    public static boolean isCompatible(String oldUrl, String newUrl);
+    public static boolean isNewer(String testUrl, String baseUrl);
+}
+```
+
+This could be used in the `Resource.merge` method to determine if resources
+can be converted to a compatible schema.  Specfically resource can be merged
+IFF:
+
+- The schemas are compatible
+- Converting all `Resource` instances to the "newest" schema version returns
+  valid `Resource` objects.
+
+## Internal details
+
+There are three major changes to SDKs for this proposal.
+
+### Understanding schema compatibility
+
+Here we propose simply providing an implementation of
+[SemVer](https://semver.org/) for isCompatible, and using the
+the [ordering rules](https://semver.org/#spec-item-11) for isNewer.  This would
+pull the version number out of the schema url.
+
+### Resource merge logic
+
+Resource merge logic is updated as follows (new mechanisms in bold):
+
+The resulting resource will have the Schema URL calculated as follows:
+
+- If the old resource's Schema URL is empty then the resulting
+  resource's Schema URL will be set to the Schema URL of the updating resource
+- Else if the updating resource's Schema URL is empty then the resulting
+  resource's Schema URL will be set to the Schema URL of the old resource,
+- Else if the Schema URLs of the old and updating resources are the same then
+  that will be the Schema URL of the resulting resource,
+- **Else if the Schema urls of the old and updating resources are compatible**
+  **then the older resource will be converted to newer schema url and merged.**
+- Else this is a merging error (this is the case when the Schema URL of the old
+  and updating resources are not empty and are different).
+
+### Resource conversion logic
+
+SDKs will be expected to provide a mechanism to apply transformations
+(both backwards and forwards) listed in `changes` for any compatible schemas.
+
+**Note: Until conversion logic is available in SDKs, any change to telemetry**
+**schema that relies on `changes` will be considered a breaking, i.e.**
+**they will lead to a major version bump.**
+
+## Trade-offs and mitigations
+
+The major drawback to this proposal is the requirement of making all schema
+migration implementations available in all SDKs.  This is mitigated by the
+following:
+
+- Initially, we could prevent any `rename` alterations to schema for compatible
+  version numbers. `convertToSchema` would check for semver compatibility and
+  just return the original resource if compatible.
+- Alternatively, we can check the newer schema for any transformations in the
+  `all` or `resource` section before failing to convert.
```

Comment:
So this would be a "poor mans" conversion function, which just figures out that the conversion is noop and can just return the original and fail otherwise, right? I believe this should cover a very large number of real-world use cases of existing resource detectors.

Reply:
Exactly.

Reply:
Would this check be limited to attributes actually present on the resources being evaluated, or would the presence of any transformation, even for non-present attributes, prevent merging/migration?